### PR TITLE
Add deferrable mode to GKEStartPodOperator

### PR DESCRIPTION
**Pull request by VladaZakharova**
Add deferrable mode to GKEStartPodOperator

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by VladaZakharova** [comment]
@potiuk @kaxil 
Hi!
Can you please check the changes for the GKEStartPod operator, please? This is the second part of the changes from the original PR. Many thanks!

File: airflow/providers/google/cloud/hooks/kubernetes_engine.py [hunk 1]
```diff
@@ -25,29 +25,44 @@
...
+    PROVIDE_PROJECT_ID,
+    GoogleBaseAsyncHook,
+    GoogleBaseHook,
+)
 
 OPERATIONAL_POLL_INTERVAL = 15
 
 
-class GKEHook(GoogleBaseHook):
+class GKEClusterHook(GoogleBaseHook):
```
Comment (kaxil) [inline airflow/providers/google/cloud/hooks/kubernetes_engine.py]: Can you add backwards-compat if possible please? Users might be using this `Hook`

Comment (VladaZakharova) [reply airflow/providers/google/cloud/hooks/kubernetes_engine.py]: Sure, good idea!
The name was changed to suite hooks' purpose: there are 2 different classes for separate hooks for cluster and pod management. I will revert the name change and leave a descriptive comment below.

**Comment by VladaZakharova** [comment]
Hi @kaxil !
Is there something else that i can improve here in the code? 
Thanks!

File: RELEASE_NOTES.rst [hunk 2]
```diff
@@ -8625,7 +8625,7 @@ Bug fixes
 - [AIRFLOW-2645][AIRFLOW-2617] Add worker_container_image_pull_policy
 - [AIRFLOW-2661] fix config dags_volume_subpath and logs_volume_subpath
 - [AIRFLOW-3550] Standardize GKE hook (#4364)
-- [AIRFLOW-2863] Fix GKEClusterHook catching wrong exception (#3711)
+- [AIRFLOW-2863] Fix GKEHook catching wrong exception (#3711)
```
Comment (kaxil) [inline RELEASE_NOTES.rst]: can you revert the unrelated change please

File: airflow/contrib/hooks/__init__.py [hunk 3]
```diff
@@ -154,7 +154,6 @@
     },
     'gcp_container_hook': {
         'GKEHook': 'airflow.providers.google.cloud.hooks.kubernetes_engine.GKEHook',
-        'GKEClusterHook': 'airflow.providers.google.cloud.hooks.kubernetes_engine.GKEHook',
```
Comment (kaxil) [inline airflow/contrib/hooks/__init__.py]: why was this entry removed?

Comment (VladaZakharova) [reply airflow/contrib/hooks/__init__.py]: My fault :)

File: RELEASE_NOTES.rst (hunk 2 above)
Comment (VladaZakharova) [reply RELEASE_NOTES.rst]: I have removed these changes

**Comment by VladaZakharova** [comment]
@kaxil
Please, check the changes so i can merge them to main

**Comment by VladaZakharova** [comment]
@potiuk 
Hi! Could you please check these changes? My team is really waiting for these changes :)

**Comment by potiuk** [comment]
I have not followed too much the deferrable operators - I know it was somewhere in between Astronomer and Goole work :) and you worked, so I guess It would take more time for me to get full context. But maybe @kaxil and @phanikumv could chime-in here?

**Comment by VladaZakharova** [comment]
@kaxil 
Hi! Can you please review my changes?

**Comment by phanikumv** [comment]
@pankajastro  and I are looking into this changes

File: airflow/providers/google/cloud/triggers/kubernetes_engine.py [hunk 4]
```diff
@@ -18,14 +18,108 @@
...
+try:
+    from airflow.providers.cncf.kubernetes.triggers.pod import KubernetesPodTrigger
+except ImportError:
+    # preserve backward compatibility for older versions of cncf.kubernetes provider
+    from airflow.providers.cncf.kubernetes.triggers.kubernetes_pod import KubernetesPodTrigger
+from airflow.providers.google.cloud.hooks.kubernetes_engine import AsyncGKEHook, AsyncGKEPodHook
 from airflow.triggers.base import BaseTrigger, TriggerEvent
 
 
+class GKEPodTrigger(KubernetesPodTrigger):
```
Comment (phanikumv) [inline airflow/providers/google/cloud/triggers/kubernetes_engine.py]: shouldnt this be `GKEStartPodTrigger` instead of `GKEPodTrigger` - its being used only by `GKEStartPodOperator`.

File: airflow/providers/google/cloud/hooks/kubernetes_engine.py [hunk 5]
```diff
@@ -336,3 +343,195 @@ async def get_operation(
...
+        :param name: Name of the pod.
+        :param namespace: Name of the pod's namespace.
+        """
+        return self.core_v1_client.read_namespaced_pod(
+            name=name,
+            namespace=namespace,
+        )
+
+
+class AsyncGKEPodHook(GoogleBaseAsyncHook):
```
Comment (phanikumv) [inline airflow/providers/google/cloud/hooks/kubernetes_engine.py]:
```suggestion
class GKEPodAsyncHook(GoogleBaseAsyncHook):
```
just to align with how Async classes are being named

File: airflow/providers/google/cloud/hooks/kubernetes_engine.py [hunk 6]
```diff
@@ -336,3 +343,195 @@ async def get_operation(
...
+        self._ssl_ca_cert = ssl_ca_cert
+
+        kwargs.update(
+            cluster_url=cluster_url,
+            ssl_ca_cert=ssl_ca_cert,
+        )
+        super().__init__(**kwargs)
+
+    @contextlib.asynccontextmanager
+    async def get_conn(self, token: Token) -> async_client.ApiClient:  # type: ignore[override]
```
Comment (phanikumv) [inline airflow/providers/google/cloud/hooks/kubernetes_engine.py:445]: why did we add `# type: ignore[override]` here

File: tests/system/providers/google/cloud/kubernetes_engine/example_kubernetes_engine_async.py [hunk 7]
```diff
@@ -68,9 +68,11 @@
         in_cluster=False,
         is_delete_operator_pod=True,
         get_logs=True,
+        deferrable=True,
     )
 
-    pod_task_xcom = GKEStartPodOperator(
+    # [START howto_operator_gke_start_pod_xcom_async]
+    pod_task_xcom_async = GKEStartPodOperator(
         task_id="pod_task_xcom",
```
Comment (phanikumv) [inline tests/system/providers/google/cloud/kubernetes_engine/example_kubernetes_engine_async.py]:
```suggestion
        task_id="pod_task_xcom_async",
```

File: airflow/providers/google/cloud/operators/kubernetes_engine.py [hunk 8]
```diff
@@ -452,89 +473,69 @@ def __init__(
         if self.config_file:
             raise AirflowException("config_file is not an allowed parameter for the GKEStartPodOperator.")
 
-    @staticmethod
-    @contextmanager
-    def get_gke_config_file(
```
Comment (pankajastro) [inline airflow/providers/google/cloud/operators/kubernetes_engine.py:457]: I believe operators/hooks public methods need to be backward compatible so not sure if we can remove it directly

File: airflow/providers/google/cloud/triggers/kubernetes_engine.py (hunk 4 above)
Comment (VladaZakharova) [reply airflow/providers/google/cloud/triggers/kubernetes_engine.py]: Done

File: airflow/providers/google/cloud/hooks/kubernetes_engine.py (hunk 5 above)
Comment (VladaZakharova) [reply airflow/providers/google/cloud/hooks/kubernetes_engine.py]: Done

File: airflow/providers/google/cloud/hooks/kubernetes_engine.py (hunk 6 above)
Comment (VladaZakharova) [reply airflow/providers/google/cloud/hooks/kubernetes_engine.py:445]: This was done to pass pre-commit check, since signature of this method is not the same as in method in super class BaseHook

File: airflow/providers/google/cloud/operators/kubernetes_engine.py [hunk 9]
```diff
@@ -452,89 +473,69 @@ def __init__(
...
+    def execute(self, context: Context):
+        """Executes process of creating pod and executing provided command inside it."""
+        self.fetch_cluster_info()
+        return super().execute(context)
 
-        with GKEStartPodOperator.get_gke_config_file(
-            gcp_conn_id=self.gcp_conn_id,
+    def fetch_cluster_info(self) -> tuple[str, str | None]:
+        """Fetches cluster info for connecting to it."""
+        cluster = self.cluster_hook.get_cluster(
```
Comment (pankajastro) [inline airflow/providers/google/cloud/operators/kubernetes_engine.py:514]: wondering if it make sense to use async `get_cluster(...)` 
https://github.com/googleapis/python-container/blob/main/google/cloud/container_v1/services/cluster_manager/async_client.py#L359

File: tests/system/providers/google/cloud/kubernetes_engine/example_kubernetes_engine_async.py (hunk 7 above)
Comment (VladaZakharova) [reply tests/system/providers/google/cloud/kubernetes_engine/example_kubernetes_engine_async.py]: Done

File: docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst [hunk 10]
```diff
@@ -148,6 +148,14 @@ And then use it in other operators:
     :start-after: [START howto_operator_gke_xcom_result]
     :end-before: [END howto_operator_gke_xcom_result]
 
+You can use deferrable mode for this action in order to run the operator asynchronously:
```
Comment (phanikumv) [inline docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst]:
```suggestion
You can set the parameter deferrable = True in order to run the operator asynchronously:
```

File: tests/providers/google/cloud/hooks/test_kubernetes_engine.py [hunk 11]
```diff
@@ -20,12 +20,13 @@
...
 from asyncio import Future
 
+import kubernetes.client
 import pytest
 from google.cloud.container_v1 import ClusterManagerAsyncClient
 from google.cloud.container_v1.types import Cluster
 
 from airflow.exceptions import AirflowException
-from airflow.providers.google.cloud.hooks.kubernetes_engine import AsyncGKEHook, GKEHook
+from airflow.providers.google.cloud.hooks.kubernetes_engine import AsyncGKEHook, AsyncGKEPodHook, GKEHook
```
Comment (phanikumv) [inline tests/providers/google/cloud/hooks/test_kubernetes_engine.py]:
```suggestion
from airflow.providers.google.cloud.hooks.kubernetes_engine import GKEAsyncHook, GKEPodAsyncHook, GKEHook
```
naming should be inline with other async hooks across the repo

File: tests/providers/google/cloud/hooks/test_kubernetes_engine.py [hunk 12]
```diff
@@ -291,6 +296,72 @@ def test_wait_for_response_running(self, time_mock, operation_mock):
         assert operation_mock.call_count == 2
 
 
+class TestAsyncGKEPodHook:
```
Comment (phanikumv) [inline tests/providers/google/cloud/hooks/test_kubernetes_engine.py]:
```suggestion
class TestGKEPodAsyncHook:
```

File: docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst (hunk 10 above)
Comment (phanikumv) [reply docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst]: Please also add how running this in deferrable mode would benefit the user

File: airflow/providers/google/cloud/operators/kubernetes_engine.py (hunk 9 above)
Comment (VladaZakharova) [reply airflow/providers/google/cloud/operators/kubernetes_engine.py:514]: There is no big need to implement here async call to get cluster since this call only used for getting endpoint to the cluster, and not perform any other long-running action on it. the async get_cluster() method is called in the trigger

File: airflow/providers/google/cloud/operators/kubernetes_engine.py (hunk 8 above)
Comment (VladaZakharova) [reply airflow/providers/google/cloud/operators/kubernetes_engine.py:457]: The method was changed to use another way of connection: instead of direct creating gcloud command to perform any action, now the ssl certificate and url of the cluster are used to connect to specific cluster. 
Can i add some `deprecation warning` for this method instead?

File: tests/providers/google/cloud/hooks/test_kubernetes_engine.py (hunk 12 above)
Comment (VladaZakharova) [reply tests/providers/google/cloud/hooks/test_kubernetes_engine.py]: Done

File: tests/providers/google/cloud/hooks/test_kubernetes_engine.py (hunk 11 above)
Comment (VladaZakharova) [reply tests/providers/google/cloud/hooks/test_kubernetes_engine.py]: Done

File: docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst (hunk 10 above)
Comment (VladaZakharova) [reply docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst]: Done

File: airflow/providers/google/cloud/operators/kubernetes_engine.py (hunk 8 above)
Comment (pankajastro) [reply airflow/providers/google/cloud/operators/kubernetes_engine.py:457]: I think good to have a warning but let's see what others think...

**Review by pankajastro — APPROVED** [review]
LGTM. But I see the removable of some public methods from the hook so maybe we needs a warning on that for backwards compatibility?

**Comment by VladaZakharova** [comment]
@pankajastro Could you please review my changes once again? Thanks! :)

**Review by potiuk — APPROVED** [review]
